packages/core/src/util/graphic.ts

```diff
@@ -325,11 +335,14 @@ const traverseUp = <T extends { children?: T[] }>(data: T, fn: (param: T) => boo
  * depth first traverse, from root to leaves, children in inverse order
  *  if the fn returns false, terminate the traverse
  */
-export const traverseTree = <T extends { children?: T[] }>(data: T, fn: (param: T) => boolean) => {
+export const traverseTree = <T extends { children?: T[] }>(
+  data: T,
+  fn: (data: T, parent: T | null, index: number) => boolean
+) => {
   if (typeof fn !== 'function') {
     return;
   }
-  traverse(data, fn);
+  traverse(data, null, -1, fn);
```

Comment:
Would `null` be a better "not in a child array" value than `-1`? `-1` does allude to stuff like `indexOf`, but `null` would prevent errors and make sure those using `index` check that it is actually a number (instead of accidentally using `-1` on an array). What do you think?

Reply:
No, it would just be a hassle, forcing you to use `as number` all the time. Basically, if a user has to guard against the "no parent" case, they will simply check on the `parent` property rather than the `index` property.

That's easier, because rather than writing `if (index < 0)` or `if (index !== null)` they just write `if (parent)`, which is nicer.

But if `index` can be either a `null` or a `number`, then you have to cast every time afterward, because Typescript wouldn't know that the two variable have "paired types"...

I think it's not worth it - `number` is fine.